pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1104,107 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
 		return worker.StatusAllOK
 	}
 
-	// cascading-delete is enabled, wait for member objects to be deleted
+	// cascading-delete is enabled, delete the member objects managed by admiral
 	ftcLister := s.ftcManager.GetFederatedTypeConfigLister()
 	ftcs, err := ftcLister.List(labels.Everything())
 	if err != nil {
-		logger.Error(err, "failed to get ftc lister")
+		logger.Error(err, "Failed to get ftc lister")
 		return worker.StatusError
 	}
 
+	// concurrent operations for each ftc
+	wg := sync.WaitGroup{}
 	remainingByGVK := make(map[string]string, len(ftcs))
-	for _, ftc := range ftcs {
-		gvk := ftc.GetSourceTypeGVK().String()
-		resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(
-			ftc.GetSourceTypeGVK(),
-			cluster.Name,
-		)
-		if !exists {
-			remainingByGVK[gvk] = fmt.Sprintf("failed to get resource lister for %s", gvk)
-			continue
-		}
+	for i := range ftcs {
+		wg.Add(1)
+		go func(ftc *fedcorev1a1.FederatedTypeConfig) {
+			defer wg.Done()
+			gvk := ftc.GetSourceTypeGVK().String()
+
+			// get all managed cluster objects according to ftc type
+			clusterObjs, namespaces, err := s.listManagedClusterObjects(ctx, ftc, cluster.Name)
+			if err != nil {
+				logger.Error(err, "Failed to list clusterObjects")
+				remainingByGVK[gvk] = fmt.Sprintf("failed to list cluster objects for %s", gvk)
+				return
+			}
+			if len(clusterObjs) == 0 {
+				return
+			}
 
-		// If cluster cache is synced, we check the store.
-		// Otherwise, we will have to issue a list request.
-		if hasSynced() {
-			objects, err := resourceLister.List(labels.Everything())
+			client, err := s.getClusterClient(cluster.Name)
 			if err != nil {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster lister: %v)", err)
-			} else if len(objects) > 0 {
-				remainingByGVK[gvk] = strconv.Itoa(len(objects))
+				logger.Error(err, "Failed to get cluster client")
+				remainingByGVK[gvk] = "failed to get cluster client"
+				return
 			}
-		} else {
-			client, exists := s.fedInformerManager.GetClusterDynamicClient(cluster.Name)
-			if !exists {
-				remainingByGVK[gvk] = "Unknown (cluster client does not exist)"
-				continue
+
+			// update some cluster objects before deleting them
+			var updateFailedObjs []string
+			for i := range clusterObjs {
+				clusterObj := clusterObjs[i].DeepCopy()
+				keyedLogger := klog.FromContext(ctx).
```

Comment:
This keyedLogger should be generated from logger rather then ctx, and the gvk infor should be included

Reply:
The code has been updated.